File: 2019/elixir/lib/intcode.ex
defmodule Intcode do
  defstruct addresses: {}, position: 0, state: :new, input: [], output: ""

  @moduledoc """
  Documentation for the Intcode computer.
  """

  def new(program, input \\ []) when is_list(program) do
    addresses = program
                |> Enum.with_index()
                |> Enum.into(Map.new, fn {n, i} -> {i, n} end)

    %__MODULE__{addresses: addresses, state: :running, input: input}
  end

  @doc """
  Hello world.

  ## Examples

      iex> %Intcode{state: :halt} = Intcode.run([99])
      iex> Intcode.run([1,0,0,0,99]) |> Map.get(:addresses) |> Map.get(0)
      2
      iex> Intcode.run([1,1,1,4,99,5,6,0,99]) |> Map.get(:addresses) |> Map.get(0)
      30
      iex> Intcode.new([3,0,4,0,99], [42]) |> Intcode.run() |> Map.get(:addresses) |> Map.get(0)
      42
      iex> Intcode.new([3,0,4,0,99], [42]) |> Intcode.run() |> Map.get(:output) |> String.trim()
      "42"

  """
  def run(program) when is_list(program) do
    new(program)
    |> run()
  end

  def run(%__MODULE__{state: :halt}=intcode), do: intcode

  def run(intcode) do
    instruction = Map.fetch!(intcode.addresses, intcode.position)
    {parameter_modes, opcode} = parse_instruction(instruction)

    do_op(opcode, intcode)
    |> run()
  end


 @doc """
  Hello world.

  ## Examples

      iex> Intcode.parse_instruction(2)
      {[:position, :position, :position], 2}
      iex> Intcode.parse_instruction(1002)
      {[:position, :immediate, :position], 2}

  """
  def parse_instruction(instruction) when is_integer(instruction) do
    <<raw_parameter_modes::binary-size(3), opcode::binary-size(2)>> =
      Integer.to_string(instruction)
      |> String.pad_leading(5, "0")

    { raw_parameter_modes
      |> String.graphemes()
      |> Enum.reverse()
      |> Enum.map(fn param_mode ->
           case param_mode do
             "0" -> :position
             "1" -> :immediate
           end
         end) ,
      opcode
      |> String.to_integer()
    }
  end

  def do_op(99, intcode) do
    %__MODULE__{intcode | state: :halt}
  end

  def do_op(1, intcode) do
    [noun_index, verb_index, write_to] = Enum.map((intcode.position+1..intcode.position+3),
                                                  fn i -> Map.fetch!(intcode.addresses, i) end)

    noun = Map.fetch!(intcode.addresses, noun_index)
    verb = Map.fetch!(intcode.addresses, verb_index)
    new_addresses = Map.put(intcode.addresses, write_to, noun + verb)
    new_position = intcode.position + 4
    %__MODULE__{intcode | addresses: new_addresses, position: new_position}
  end

  def do_op(2, intcode) do
    [noun_index, verb_index, write_to] = Enum.map((intcode.position+1..intcode.position+3),
                                                  fn i -> Map.fetch!(intcode.addresses, i) end)

    noun = Map.fetch!(intcode.addresses, noun_index)
    verb = Map.fetch!(intcode.addresses, verb_index)
    new_addresses = Map.put(intcode.addresses, write_to, noun * verb)
    new_position = intcode.position + 4
    %__MODULE__{intcode | addresses: new_addresses, position: new_position}
  end

  def do_op(3, intcode) do
    write_to = Map.fetch!(intcode.addresses, intcode.position+1)
    [input | rest] = intcode.input
    new_addresses = Map.put(intcode.addresses, write_to, input)
    new_position = intcode.position + 2
    %__MODULE__{intcode | addresses: new_addresses, input: rest, position: new_position}
  end

  def do_op(4, intcode) do
    output_position = Map.fetch!(intcode.addresses, intcode.position+1)
    output = Map.fetch!(intcode.addresses, output_position)
    updated_output = intcode.output <> to_string(output) <> "\n"
    new_position = intcode.position + 2
    %__MODULE__{intcode | output: updated_output, position: new_position}
  end
end

File: 2019/ruby/lib/intcode.rb
class Intcode
  attr_reader :name, :idx, :program, :debug, :states
  attr_accessor :memory, :input, :output, :relative_base

  def initialize(idx: 0, program: [99], input: [], debug: false)
    @idx = idx
    @name = ('A'..'Z').to_a[idx]
    @program = program
    @input = input
    @debug = debug

    @memory = program.dup
    @relative_base = 0
    @output = []
    @status = :initializing
    @states = [current_state]
    debug_puts "[#{name}] 🐣  #{input.to_s}"
  end

  def current_state
    { status: @status,
      memory: @memory.dup,
      pointer: @pointer.dup,
      relative_base: @relative_base.dup,
      params: @params.dup,
      input: @input.dup,
      #output: @output.dup,
    }
  end

  def run
    pointer = 0
    @status = :starting
    @states << self.current_state
    debug_puts "[#{name}] 🟢"
    loop do
      instruction = Instruction.new(memory[pointer])

      params = memory[pointer+1..pointer+instruction.arity]

      pointer_action, location = instruction.execute(self, params)
      case pointer_action
      when :jump
        pointer = location
      when :advance
        pointer += location
      when :halt
        @status = :halted
        @states << self.current_state
        debug_puts "[#{name}] 🛑"
        return memory
      end
      @status = :running
      @states << self.current_state.merge(instruction: instruction.to_s)
    end
  rescue Instruction::InvalidOpcode => e
    debug ? binding.pry : raise(e)
  end

  def status
    states.last[:status]
  end

  def halted?
    status == :halted
  end

  def read_input
    sleep_iter = 0
    while @input.first.nil?
      sleep_iter += 1
      debug_puts "[#{@name}] is waiting for input ..." if sleep_iter % 1000 == 0
      sleep 0.01
    end
    input = @input.shift
    debug_puts "[#{@name}] 👁  #{input}"
    raise "🎅 HO HO oh no! 💥" if input.nil?
    input
  end

  def receive_input(values)
    debug_puts "[#{@name}] 📨  #{values.inspect}"
    if values.kind_of? Array
      values.each {|v| @input << v }
    else
      @input << values
    end
  end

  def send_output(value)
    debug_puts "[#{@name}] 📬  #{value}"
    @output << value
  end

  def debug_puts(msg)
    puts msg if debug
  end
end

class Instruction
  attr_reader :arity

  class InvalidOpcode < StandardError; end

  Op = Struct.new(:name, :go) do
    def arity
      go.parameters.length - 1
    end

    def location_params_mask
      go.parameters.map do |_, param_sym|
        param_sym.to_s.end_with?("_location")
      end[1..]
    end
  end

  INSTRUCTIONS = {
    1 => Op.new(:add, -> (state, noun, verb, write_location) { state.memory[write_location] = noun + verb ; :advance } ),
    2 => Op.new(:multiply, -> (state, noun, verb, write_location) { state.memory[write_location] = noun * verb ; :advance } ),
    3 => Op.new(:input, -> (state, write_location) { state.memory[write_location] = state.read_input.to_i ; :advance } ),
    4 => Op.new(:output, -> (state, output) { state.send_output(output) ; :advance } ),
    5 => Op.new(:jump_if_true, -> (_state, jump_flag, jump_to) { jump_flag == 0 ? :advance : jump_to } ),
    6 => Op.new(:jump_if_false, -> (_state, jump_flag, jump_to) { jump_flag == 0 ? jump_to : :advance } ),
    7 => Op.new(:less_than, -> (state, noun, verb, write_location) { state.memory[write_location] = noun < verb ? 1 : 0 ; :advance } ),
    8 => Op.new(:equals, -> (state, noun, verb, write_location) { state.memory[write_location] = noun == verb ? 1 : 0 ; :advance } ),
    9 => Op.new(:update_relative_base, -> (state, base_offset) { state.relative_base += base_offset ; :advance }),
    99 => Op.new(:halt, ->(_state) { :halt } ),
  }
  VALID_OPCODES = INSTRUCTIONS.keys

  def initialize(raw_instruction)
    @raw_instruction = raw_instruction
    unless VALID_OPCODES.include? opcode
      raise InvalidOpcode, "#{opcode} is not a valid opcode. Valid: #{VALID_OPCODES.join(", ")}. Raw instruction was #{@raw_instruction}."
    end
    @instruction = INSTRUCTIONS[opcode]
    @arity = @instruction.arity
    @orig_params = []
    @moded_params = []
  end

  def opcode
    @opcode ||= parsed_instruction[-2..-1].join.to_i
  end

  PARAMETER_MODE_MAP = {
    "0" => :position,
    "1" => :immediate,
    "2" => :relative
  }

  def parameter_modes
    @parameter_modes ||= parsed_instruction.take(3)
                                           .reverse
                                           .zip(@instruction.location_params_mask)
                                           .map do |mode_digit, location_param|
                                             if location_param
                                               (PARAMETER_MODE_MAP[mode_digit].to_s + "_location").to_sym
                                             else
                                               PARAMETER_MODE_MAP[mode_digit]
                                             end
                                           end
  end

  def parsed_instruction
    @parsed_instruction ||= @raw_instruction.to_s.rjust(5, "0").chars
  end

  def execute(state, params)
    @orig_params = params
    @moded_params = apply_modes(state, params)
    result = @instruction.go.call(state, *@moded_params)
    case result
    when :advance
      [:advance, arity + 1]
    when :halt
      [:halt, 0]
    else
      [:jump, result]
    end
  rescue TypeError
    binding.pry
  end

  def apply_modes(state, params)
    params.zip(parameter_modes).map do |param, mode|
      case mode
      when :position
        state.memory.fetch(param, 0)
      when :position_location
        param
      when :relative
        state.memory.fetch(param + state.relative_base, 0)
      when :relative_location
        param + state.relative_base
      when :immediate
        param
      else
        raise "What kind of weird mode are you in?"
      end
    end
  end

  def to_h
    { raw_instruction: @raw_instruction,
      opcode: opcode,
      name: @instruction.name,
      parameter_modes: parameter_modes,
      orig_params: @orig_params,
      moded_params: @moded_params,
    }
  end

  def to_s
    to_h.to_s
  end
end
